### DISCUSSION TITLE
Why method of AppHeader component is not triggered on emit?

### DISCUSSION THREAD

**Discussion by sergeynilov** [livewire/livewire#3999] [1. Help]
In Laravel 8, livewire2 I make a header component , which I call in 
resources/views/layouts/frontpage.blade.php :

    </head>
    
    <body class="flex flex-col min-h-screen">
    <header>
        @livewire('app-header', ['layout'=>'frontend'])
    </header>
    
    <main class="frontend_page_container_wrapper z-20 ">
        {{ $slot }}
    </main>
    
    <footer class="bg-green" style="height: 138px !important;">
        @livewire('frontend-footer')
    </footer>
    
    @stack('modals')
    
    @livewireScripts
    </body>
    
    </html>

To update current info in header when new page is opened in any component I call event of AppHeader component.
Like in app/Http/Livewire/Hostel/HostelsHomepageSpotlight.php :

    <?php
    
    namespace App\Http\Livewire\Hostel;
    
    ...
    use Livewire\Component;
    
    class HostelsHomepageSpotlight extends Component
    {
        
        ...
        public function render()
        {
            return view('livewire.hostel.hostels-homepage-spotlight', [
            ])->layout('layouts.frontpage');
        } // public function render()
        
        public function mount()
        {
            
            $this->hostelsDataRows = Hostel
                ::getByStatus('A')
                ...
                ->paginate($hostels_per_page);
                ...
            
            \Log::info(  varDump(-1, ' -1 HostelsHomepageSpotlight before setCurrentPageProperties::') );
    //        $this->emitUp('setCurrentPageProperties', [  // IF TO UNCOMMENT THIS LINE - IT DOES NOT WORK
            $this->emitTo('AppHeaderAppHeader','setCurrentPageProperties', [
                'layout'                => "frontend",
                'icon'                  => 'hostel',
                'additive_class'        => '',
                'title'                 => 'Spotlight hostels',
                'trigger_js_event'      => true,
                'template_container_id' => "hostels_homepage_spotlight_page_container"
            ]);
            
        } // public function mount()
        ...

and in app/Http/Livewire/AppHeader.php :

    <?php
    
    namespace App\Http\Livewire;
    
    use Livewire\Component;
    
    class AppHeader extends Component
    {
        ...
        protected $listeners = ['setCurrentPageProperties' => 'setCurrentPagePropertiesCallback'];
        public function render()
        {    
            return view('livewire.app-header',  [])
                ->layout('layouts.') ;//. $this->layout;
        }
        
        public function mount(
            ...
        ) {
            ...
        }
        
        
        public function setCurrentPagePropertiesCallback($data)
        {
            \Log::info(varDump($data, 'INSIDE  -1  setCurrentPagePropertiesCallback $data::'));
            if ( ! empty($data['layout'])) {
                $this->layout = $data['layout'];
            }
            ...
        } // public function setCurrentPagePropertiesCallback($data)
        
    }

When I open page with HostelsHomepageSpotlight.php component
I see degugging line 

    -1 HostelsHomepageSpotlight before setCurrentPageProperties

but not 

    INSIDE  -1  setCurrentPagePropertiesCallback 

as setCurrentPagePropertiesCallback method of HostelsHomepageSpotlight.php is not triggered.

How can it be fixed?

Thanks!

**Comment by jonneroelofs** [comment]
Have you tried using emit instead of emitUp?

**Comment by joshhanley** [comment]
@jonneroelofs is correct, you can't use `emitUp` in this scenario as one component isn't nested inside the other component. So your best to keep using `emitTo` or just use `emit` as suggested above.

Hope this helps!

**Reply by sergeynilov** [reply]
I had misspelling in name of target component. It is “AppHeader”, so fixing  :

`            $this->emitTo('AppHeader','setCurrentPageProperties', [
`
this event is not triggered in app/Http/Livewire/AppHeader.php, with declared :
```
    class AppHeader extends Component
    {  
        ...
        protected $listeners = ['setCurrentPageProperties' => 'setCurrentPagePropertiesCallback'];
        ...
        public function setCurrentPagePropertiesCallback($data)
        {
            \Log::info(varDump($data, 'INSIDE  -1  setCurrentPagePropertiesCallback $data::'));
        ...
        }
    }        

```        

I also tried :
`            $this->emit('setCurrentPageProperties', [
`But it does not work anyway...

**Comment by sergeynilov** [comment]
In AppHeader Component resources/views/livewire/app-header.blade.php I added calling method of AppHeader Component  as :
```
    <a  wire:click="appHeaderClickTest"  class=" block text-sm px-2 py-4 hover:bg-green-500 transition duration-300 ">
       Test2
    </a>

```
with defintion in app/Http/Livewire/AppHeader.php :
```
    public function appHeaderClickTest() {
        \Log::info(  varDump(-12, ' -12 appHeaderClickTest::') );
    }

```
and clicking on "Test2" button I see message in log file.
So I check that I put valid AppHeader Component and it works ok when I click "Test2" button

**Comment by sergeynilov** [comment]
Looking at the code of resources/views/layouts/frontpage.blade.php again, I see that
 @livewire('app-header', is called in block  which is out of {{ $slot }}. Can it be the issue 
 and how can it be fixed then ?

**Reply by joshhanley** [reply]
@sergeynilov that shouldn't be a problem, you can have Livewire component in other places in your app layout.

I'm not exactly sure what is going on, as that setup works fine for me.

Could you share your latest code for your two component classes and views, as well as your layout and route definition, so we can see the full picture?

Thanks!

**Comment by sergeynilov** [comment]
I created new livewire app EventsTest

with 4 components:

php artisan make:livewire Home
php artisan make:livewire hostel/HostelsHomepageSpotlight
php artisan make:livewire AppHeader
php artisan make:livewire hostel/HostelViewPage

and I placed HostelsHomepageSpotlight on home page.
In my app just 2 links and AppHeader at header of resources/views/layouts/app.blade.php
I emit event setCurrentPageProperties but it is not triggered in AppHeader (I have default title and no logs rows)

I uploaded it on https://github.com/sergeynilov/EventsTest
Could you please to look at it?

Thanks!

**Comment by joshhanley** [comment]
@sergeynilov finally going through my backlog and just had a look at this.

A couple of things to note, one is, that events can't be emitted on the first request in mount or render, as Livewire's javascript isn't running yet. So changing to the `wire:init` like you have done is the best way.

I think there was a syntax error in `AppHeader::setCurrentPagePropertiesCallback()` as when I comment out the first line `\Log::info...`, then the titles started to change as expected.

![image](https://user-images.githubusercontent.com/882837/142500484-4a89b4a7-5983-405b-9276-ece330f3b66b.png)

Hope this helps!
